### ISSUE TITLE
Does librariesio's source rank's definition of SemVer accept post release versions?

### ISSUE THREAD

**Issue by phseiff** [librariesio/libraries.io#2771]
Hello,
I have some questions about the way the Source Rank is calculated, that weren't solved by the documentation.

I am thinking about creating a post-release version for one of my projects (titled `1.0.0-post1`). This version would be the first one released after version `1.0.0`.
According to [the SemVer specification](https://semver.org/), this is a valid SemVer version number. However, the SemVer specification only talks about version numbers like this in the context of pro-releases. Would the Source Rank still accept it as semantic versioning if I released the `1.0.0-post1` after the `1.0.0`? The SemVer specification is slightly unspecific about that, so I would say that the Source Rank should probably err on the side of the maintainer on this, especially since using post-releases to fix smaller things in a projects README, such as spelling mistakes in a changelog, appears to be common practice in SemVer-versioned open source projects.
I read the [distribution-section](https://docs.libraries.io/overview#distribution) in the Source Rank section of the documentation, and it does not really answer that question.

Another question I have is whether the logarithm that is calculated for some Source Rank attributes (mainly in the [community-section](https://docs.libraries.io/overview#community) of the Source Rank) is rounded up or down after calculation, and whether the logarithm's base is 2 (which is what I expected), or something else.

My apologies in case I should have used the Issue template for this, but it didn't have a section for questions, and I wasn't sure if I should use the *Feature Suggestion* section for it or not.

**Comment by phseiff** [comment]
I think I can answer my own question new (I would still appreciate if someone could tell me whether I am right with it, though).

I just took a look at the code base, and it seems to be that it validates the version numbers version for version rather than looking at their order. For each version, the result of [`SemanticRange.clean(version)`](https://github.com/librariesio/semantic_range/#usage) is calculated (see [here](https://github.com/librariesio/libraries.io/blob/6aeb4640777b62b6128a25b745dc6c2615e667bb/app/models/concerns/releaseable.rb#L40) in the code base) and fed into [`Semantic::Version.new`](https://github.com/jlindsey/semantic/#usage); the version is then considered to be a valid version number if said function doesn't fail with an `ArgumentError` (see [here](https://github.com/librariesio/libraries.io/blob/6aeb4640777b62b6128a25b745dc6c2615e667bb/app/models/concerns/releaseable.rb#L7) in the code base).
If all versions of the package have valid SemVer-versions, the package is considered to use valid SemVer, if I understand it correctly (see [here](https://github.com/librariesio/libraries.io/blob/6aeb4640777b62b6128a25b745dc6c2615e667bb/app/models/concerns/source_rank.rb#L64) in the code base).
Since `Semantic::Version.new` accepts e.g. `1.6.5-post0` as a valid SemVer-version, using such versions on PyPi should be okay without decreasing the source rank, if I understand it correctly.

I am slightly confused, though, since there appears to be a second implementation [here](https://github.com/librariesio/libraries.io/blob/6aeb4640777b62b6128a25b745dc6c2615e667bb/app/models/project_score_calculator.rb#L134) that only looks at the last ten versions of the project (which seems to differ from the current behavior of libraries.io). `app/models/source_rank.rb` and `app/models/project_score_calculator.rb` appear to do the same thing in different ways, if I understand it correctly, so which one is the one that's actually being used?

> Another question I have is whether the logarithm that is calculated for some Source Rank attributes (mainly in the community-section of the Source Rank) is rounded up or down after calculation, and whether the logarithm's base is 2 (which is what I expected), or something else.

I looked at the code in `app/models/project_score_calculator.rb`, and to my understanding, it uses the `log10`. My knowledge of Ruby does not suffice to judge whether this is rounded down or up afterwards, though.

TL;DR:

`libraries.io` accepts `1.2.3-post0` as a valid SemVer-version (at least to my understanding), and the logarithms are calculated with basis 10 (to my understanding).

**Comment by phseiff** [comment]
> I am slightly confused, though, since there appears to be a second implementation here that only looks at the last ten versions of the project (which seems to differ from the current behavior of libraries.io). `app/models/concerns/source_rank.rb` and `app/models/project_score_calculator.rb` appear to do the same thing in different ways, if I understand it correctly, so which one is the one that's actually being used?

I just re-checked, and it appears I was looking at an older version of the project (some commits back) before. The semantic versioning calculation code is not part of `app/models/concerns/source_rank.rb` anymore; my apologies for not realizing that earlier.
